File: harness/fuzz/runner/campaigns.py
from __future__ import annotations

import hashlib
from dataclasses import dataclass


CAMPAIGN_CHOICES: tuple[str, ...] = (
    "none",
    "ram-bitflip",
    "reg-bitflip",
)


@dataclass(frozen=True)
class Campaign:
    name: str
    fault_mode: str
    fault_domain: str
    requires_injection: bool = False
    requires_fuzz_symbols: bool = False


CAMPAIGNS: dict[str, Campaign] = {
    "none": Campaign("none", "none", "none"),
    "ram-bitflip": Campaign(
        "ram-bitflip",
        "ram-bitflip",
        "ram",
        requires_injection=True,
        requires_fuzz_symbols=True,
    ),
    "reg-bitflip": Campaign(
        "reg-bitflip",
        "reg-bitflip",
        "register",
        requires_injection=True,
    ),
}


def campaign(name: str) -> Campaign:
    try:
        return CAMPAIGNS[name]
    except KeyError as exc:
        choices = ", ".join(CAMPAIGN_CHOICES)
        raise ValueError(
            f"unsupported campaign {name!r}; expected one of: {choices}") from exc


def derive_trial_seed(
    *,
    campaign_seed: int,
    trial_id: int,
    technique: str,
    implementation: str,
    campaign_name: str,
) -> int:
    payload = (
        f"{campaign_seed:016x}:{trial_id}:{technique}:"
        f"{implementation}:{campaign_name}"
    ).encode("utf-8")
    digest = hashlib.blake2b(payload, digest_size=8,
                             person=b"ft-single").digest()
    seed = int.from_bytes(digest, "little")
    return seed or 0x9E3779B97F4A7C15

File: harness/tui/internal/fuzz/campaigns.go
// Package fuzz ports harness/fuzz/runner: single-shot QEMU plugin fuzz
// campaigns with per-trial manifests and the 9-way result classifier.
package fuzz

import (
	"encoding/binary"
	"fmt"
	"strings"

	"github.com/dchest/blake2b"
)

type Campaign struct {
	Name                 string
	FaultMode            string
	FaultDomain          string
	RequiresInjection    bool
	RequiresFuzzSymbols  bool
	RequiresOneInsnPerTB bool
}

var CampaignChoices = []string{"none", "ram-bitflip", "reg-bitflip", "insn-skip"}

var campaigns = map[string]Campaign{
	"none": {Name: "none", FaultMode: "none", FaultDomain: "none"},
	"ram-bitflip": {
		Name:                "ram-bitflip",
		FaultMode:           "ram-bitflip",
		FaultDomain:         "ram",
		RequiresInjection:   true,
		RequiresFuzzSymbols: true,
	},
	"reg-bitflip": {
		Name:              "reg-bitflip",
		FaultMode:         "reg-bitflip",
		FaultDomain:       "register",
		RequiresInjection: true,
	},
	"insn-skip": {
		Name:                 "insn-skip",
		FaultMode:            "insn-skip",
		FaultDomain:          "instruction",
		RequiresInjection:    true,
		RequiresOneInsnPerTB: true,
	},
}

// CampaignByName mirrors campaigns.campaign.
func CampaignByName(name string) (Campaign, error) {
	spec, ok := campaigns[name]
	if !ok {
		return Campaign{}, fmt.Errorf("unsupported campaign %q; expected one of: %s",
			name, strings.Join(CampaignChoices, ", "))
	}
	return spec, nil
}

// trialSeedFallback replaces an all-zero digest, like the Python
// `seed or 0x9E3779B97F4A7C15`.
const trialSeedFallback = 0x9E3779B97F4A7C15

// DeriveTrialSeed mirrors campaigns.derive_trial_seed: BLAKE2b with an
// 8-byte digest and the "ft-single" personalization, read little-endian.
// x/crypto/blake2b cannot express the personalization — see PLAN.md §3.
func DeriveTrialSeed(campaignSeed uint64, trialID int, technique, implementation, campaignName string) uint64 {
	payload := fmt.Sprintf("%016x:%d:%s:%s:%s",
		campaignSeed, trialID, technique, implementation, campaignName)
	hash, err := blake2b.New(&blake2b.Config{
		Size:   8,
		Person: []byte("ft-single"),
	})
	if err != nil {
		panic(err) // static config; cannot fail
	}
	hash.Write([]byte(payload))
	seed := binary.LittleEndian.Uint64(hash.Sum(nil))
	if seed == 0 {
		return trialSeedFallback
	}
	return seed
}
